### DISCUSSION TITLE
Is it possible to make i3 focus on a window class that is NOT certain instances?

### DISCUSSION THREAD

**Discussion by Samueru-sama** [i3/i3#5919] [Q&A] [answered]
I'm using PWAs on the brave browser, and I have this script mapped to F4 for launching the applications, and if they are already opened it will instead focus on them: 

```
#!/bin/sh

( i3-msg '[class="Brave" instance="web.telegram.org__k"] focus' && dunstify -r 38 -t 1500 "Telegram" ) ||
( i3-msg '[class="Brave" instance="discord.com__app"] focus' && dunstify -r 38 -t 1500 "Discord" ) ||
( i3-msg '[class="Brave" instance="web.whatsapp.com"] focus' && dunstify -r 38 -t 1500 "WhatsApp" )

if [ $? -ne 0 ]; then
    dunstify -r 38 -t 2000 "Launching Chat Apps" &
    bravetelegram &
    bravediscord &
    bravewhatsapp &
    counter=0
    while ! i3-msg '[class="Brave" instance="web.telegram.org__k"] focus'; do
        sleep 0.3
        ((counter++))
        if [ "$counter" -ge 30 ]; then
            break
        fi
    done
    i3-msg layout tabbed &
fi
```

My issue is that I also have a similar script for the main browser window, however that window has an instance that constantly changes name so I can tell i3-msg to focus on the Brave class alone, because it will sometimes focus on the PWAs instances instead.

So I need to find a way to tell i3-msg to focus on a window class that is NOT one of the instances mentioned above. 

Btw: If you wonder why I have a that weird while loop in the script it is to work around a bug that causes the gaps around the windows to increase if the layout is set to tabbed before the windows are shown.

**Comment by orestisfl** [comment, marked as answer]
Yes, there is an example in the userguide:

```
# kill all windows except for Firefox and Gnome Terminal.
# case-insensitive and uses negative lookaheads, supported by PCRE
bindsym $mod+x [class="^(?i)(?!firefox)(?!gnome-terminal).*"] kill
```

just adjust it for `instance`.

**Reply by Samueru-sama** [reply]
Thank you, that worked.

`bindsym $mod+F1 exec --no-startup-id pgrep brave && i3-msg [class="Brave" instance="^(?i)(?!web.telegram.org__k)(?!discord.com__app)(?!web.whatsapp.com).*"] focus && dunstify -r 33 -t 1500 "Brave" || ( brave & dunstify -r 38 -t 2000 "Launching Brave" )`

For some reason if I try to run the `i3-msg [class="Brave" instance="^(?i)(?!web.telegram.org__k)(?!discord.com__app)(?!web.whatsapp.com).*"] focus` on the terminal when testing it both bash and zsh complain that web.telegram is not found?! but it works when put in the i3 config. No idea what is happening there.

**Reply by orestisfl** [reply]
Yes, you need to work around your shell interpreting the syntax:

```shell
i3-msg '[class="Brave" instance="^(?i)(?!web.telegram.org__k)(?!discord.com__app)(?!web.whatsapp.com).*"] focus'
```
